File: Software/Code/iMain/comms/iServer.c


#include "iServer.h"

static int THREAD_MODE = 0;            ///multi-thread mode status.

void startServer() {
   pthread_mutex_init(&server_mutex, NULL); //probably do not need server mutex
   initializeServer();

   THREAD_MODE = 1;
   pthread_mutex_init(&status_cache_mutex, NULL);
   //sensor_cache = (status_cache*) malloc(sizeof(status_cache)); //initialize status cache?
   pthread_create( &server_thread, NULL, (void*)serverThreadFunc, NULL);
   printf("Started Server Thread\n");
}

void initializeServer()
{
  struct in_addr mcastAddr;
  struct hostent *h;
  sock = 0; 
  rc = 0;
  cliLen = 0;
  memset(read_buff, 0, 54);

  /* get mcast address to listen to */
  h=gethostbyname(CREATE_GROUP);
  if(h==NULL) {
    printf("iServer : unknown group '%s'\n",CREATE_GROUP);
    exit(1);
  }
  
  memcpy(&mcastAddr, h->h_addr_list[0],h->h_length);
  
  /* check given address is multicast */
  if(!IN_MULTICAST(ntohl(mcastAddr.s_addr))) {
    printf("iServer : given address '%s' is not multicast\n",
	   inet_ntoa(mcastAddr));
    exit(1);
  }

  /* create socket */
  sock = socket(AF_INET,SOCK_DGRAM,0);
  if(sock<0) {
    printf("iServer : cannot create socket\n");
    exit(1);
  }

  /* bind port */
  servAddr.sin_family=AF_INET;
  servAddr.sin_addr.s_addr=htonl(INADDR_ANY);
  servAddr.sin_port=htons(SERVER_PORT);  
  if(bind(sock,(struct sockaddr *) &servAddr, sizeof(servAddr))<0) {
    printf("iServer : cannot bind port %d \n",SERVER_PORT);
    exit(1);
  }

  /* Get wlan0 address */
  int fd = socket(AF_INET, SOCK_DGRAM, 0);
  ifr.ifr_addr.sa_family = AF_INET;
  strncpy(ifr.ifr_name, "wlan0", IFNAMSIZ-1);
  ioctl(fd, SIOCGIFADDR, &ifr);
  //printf("%s\n", inet_ntoa(((struct sockaddr_in *)&ifr.ifr_addr)->sin_addr));

  /* join multicast group */
  mreq.imr_multiaddr.s_addr=mcastAddr.s_addr;
  mreq.imr_interface.s_addr= inet_addr(inet_ntoa(((struct sockaddr_in *)&ifr.ifr_addr)->sin_addr)); //wlan0 address

  rc = setsockopt(sock,IPPROTO_IP,IP_ADD_MEMBERSHIP,
		  (void *) &mreq, sizeof(mreq));
  if(rc<0) {
    printf("iServer : cannot join multicast group '%s'\n",
	   inet_ntoa(mcastAddr));
    exit(1);
  }
  else {
    printf("iServer : listening to mgroup %s:%d\n",
	   inet_ntoa(mcastAddr), SERVER_PORT);
 }

 not_done = 1;
}

/** Thread responsible for handling sensor loop in multi-threaded mode.
 *
 */
void *serverThreadFunc( void *ptr )
{

   fd_set readReadySet;
   
    /* Set a 5 second time-out on the select */
    struct timeval timeOut;
    timeOut.tv_sec = 5L; //5 second time out
    timeOut.tv_usec = 0L;

    no_data = 1;

    /* infinite server loop */
    while(not_done) {
        FD_ZERO(&readReadySet);
        FD_SET(sock, &readReadySet);
        timeOut.tv_sec = 5L;
        timeOut.tv_usec = 0L;
	if(select(sock+1, &readReadySet, NULL, NULL, &timeOut) > 0) {
	   if (FD_ISSET(sock, &readReadySet))
   	   {
		no_data = 0;
     		cliLen=sizeof(cliAddr);
      		//int n = recvfrom(sock,(create_status *) &status,sizeof(status),0,(struct sockaddr *) &cliAddr, &cliLen);
		int n = recvfrom(sock, read_buff, 54, 0, (struct sockaddr *) &cliAddr, &cliLen);
      		if(n<0) {
			printf("iServer : cannot receive data\n");
			continue;
      		}
		if(inet_ntoa(cliAddr.sin_addr) == inet_ntoa(((struct sockaddr_in *)&ifr.ifr_addr)->sin_addr))
			continue;
	pthread_mutex_lock( &status_cache_mutex);
	/*printf("iServer : received data from %s:%d on %s : %s at Pos[%d,%d,%d]\n",
	     inet_ntoa(cliAddr.sin_addr),ntohs(cliAddr.sin_port),
	     CREATE_GROUP,
	     status.uid, status.pos_x, status.pos_y, status.heading);*/
	int msg_type = 0;
	memcpy(&msg_type, read_buff, 1);
	switch(msg_type) {
	case MSG_TYPE_STATUS:
		memcpy((create_status *) &status, read_buff+1, sizeof(status));
		//add status to status cache map
        	put(&map_status, status.uid,status);
		break;
	case MSG_TYPE_TERMINATION:
		memcpy((asset_terminated *) &assetTerminated, read_buff+1, sizeof(assetTerminated));
		//check which asset was terminated
		break;
	case MSG_TYPE_DETECTION:
		memcpy((radar_detection *) &radarDetection, read_buff+1, sizeof(radarDetection));
		//check which asset was detected
		break;
        default:
		printf("iServer: unknown message type received\n");
		break;
	}
	memset(read_buff, 0, 54);
	pthread_mutex_unlock( &status_cache_mutex);
   	   } else {
		//printf("select timeout\n");
		continue; //timeout
           }
	}
    }/* end of infinite server loop */

    pthread_exit(NULL);
}

void closeServer ()
{      
   pthread_mutex_lock( &status_cache_mutex );
   not_done = 0;
   close(rc);
   rc = 0;
   shutdown(sock,SHUT_RDWR);
   sock = 0;
   pthread_mutex_unlock( &status_cache_mutex );
       
   pthread_join(server_thread, NULL);
   printf("Exiting Server...\n");
   pthread_mutex_destroy(& status_cache_mutex);
   pthread_mutex_destroy(&server_mutex);

   //TODO free any memory here
}

create_status* getStatus(char* uid)
{
   return find(&map_status, uid);
}

void put(Map *map, char* key, create_status value) {
   int index = findIndex(map, key);
   if(index<0)
	return;
   memcpy((char *) &map->list[index].key, key,strlen(key));
   memcpy((create_status*) &map->list[index].value, (create_status*) &value, sizeof(value));
}

static int compare(const void * p1, const void * p2)
{
    return strcmp(*((const char **)p1), *((const char **)p2));
}

int findIndex(Map *map, char* key) {
    struct item * item = bsearch(&key, robots, 10, sizeof(*robots),
        compare);

    return item ? item->value : -1;
}

create_status* find(Map* map, char* key) {
  int index = findIndex(map, key);
  if(index>=0)
	return &(map->list[index].value);
  return NULL;
}

File: Software/java/robot-display/src/com/jaws/comms/iServer.java

// Import some needed classes
package com.jaws.comms;

import java.io.ByteArrayInputStream;
import java.io.IOException;
import java.net.DatagramPacket;
import java.net.InetAddress;
import java.net.InetSocketAddress;
import java.net.MulticastSocket;
import java.net.NetworkInterface;
import java.net.UnknownHostException;
import java.util.HashMap;
import java.util.Map;

import com.jaws.display.ArenaDisplay.Listener;


public class iServer implements Runnable {
	
	
	public static final int MSG_TYPE_STATUS = 1;
	
	public static final int MSG_TYPE_TERMINATION = 2;
	
	public static final int MSG_TYPE_DETECTION = 3;
	
	/***/
	protected static final int CREATE_PORT = 1500;
	
	/***/
	protected static final String CREATE_GROUP = "225.0.0.37";
	
	/***/
	protected MulticastSocket s;
	
	protected boolean not_done;
	
	protected Map<String, createStatus> map_status = 
			new HashMap<String, createStatus>();
	
	protected Listener listener;

	public iServer(Listener listener) {
		not_done = true;
		this.listener = listener;
	}
	
	public void initializeServer() throws IOException {
	    // Create the socket and bind it to port 'port'.
		System.out.println("Initializing socket...");
	    s = new MulticastSocket(CREATE_PORT);
	    
	    // join the multicast group
	    System.out.println("Joining multicast group...");
	    
	    InetSocketAddress socketAddress = new InetSocketAddress(CREATE_GROUP, CREATE_PORT);
	    NetworkInterface ni = NetworkInterface.getByName("wlan0");
	    s.joinGroup(socketAddress,ni);
	    
	    System.out.println("Listening for data...");
	    (new Thread(this)).start(); //TODO confirm this is correct
	}
	
	protected void listen() throws IOException {
		while(not_done) {
		    // Create a DatagramPacket and do a receive
		    //createStatus status = new createStatus();
		    //DatagramPacket pack = new DatagramPacket(status, sizeof(status));
			byte buf[] = new byte[1024];
			DatagramPacket packet = new DatagramPacket(buf, buf.length);
			s.receive(packet);
			ByteArrayInputStream b_in = new ByteArrayInputStream(
					packet.getData());
			//check for message type
			byte mt[] = new byte[1];
			b_in.read(mt, 0, 1);
			int msg_type = (int)mt[0];
			switch (msg_type) {
			case MSG_TYPE_STATUS:
				createStatus stat = readStatus(b_in);
				map_status.put(stat.getUID(), stat);
				listener.recv(stat);
				System.out.println(stat.uid + " POS: [" + stat.pos_x + "," + stat.pos_y
						+ "]");
				break;
			case MSG_TYPE_TERMINATION:
				assetTerminated at = readTermination(b_in);
				System.out.println(at.getUID() + " terminated.");
				break;
			case MSG_TYPE_DETECTION:
				radarDetection rd = readDetection(b_in);
				System.out.println(rd.getUID() + " detected by radar.");
				break;
			default:
				System.out.println("Unknown message type received");
				break;
			}
			
			System.out.println("Received data from: "
					+ packet.getAddress().toString() + ":" + packet.getPort()
					+ " with length: " + packet.getLength());
		}
	}
	
	public void run() {
		try {
			listen();
		} catch (IOException e) {
			// TODO Auto-generated catch block
			e.printStackTrace();
		}
	}
	
	/**
	 * 
	 * @param b_in
	 * @return
	 */
	protected createStatus readStatus(ByteArrayInputStream b_in) {
		createStatus stat = new createStatus();
		byte buffer[] = new byte[4];
		b_in.read(buffer, 0, 4);
		stat.pos_x = (buffer[3] << 24 | (buffer[2] & 0xFF) << 16 | (buffer[1] & 0xFF) << 8 | (buffer[0] & 0xFF));
		b_in.read(buffer, 0, 4);
		stat.pos_y = (buffer[3] << 24 | (buffer[2] & 0xFF) << 16 | (buffer[1] & 0xFF) << 8 | (buffer[0] & 0xFF));
		b_in.read(buffer, 0, 4);
		//heading
		stat.heading = (buffer[3] << 24 | (buffer[2] & 0xFF) << 16 | (buffer[1] & 0xFF) << 8 | (buffer[0] & 0xFF));
		//velocity
		b_in.read(buffer, 0, 4);
		stat.velocity = (buffer[3] << 24 | (buffer[2] & 0xFF) << 16 | (buffer[1] & 0xFF) << 8 | (buffer[0] & 0xFF));
		//catagory
		b_in.read(buffer, 0, 4);
		stat.category = (buffer[3] << 24 | (buffer[2] & 0xFF) << 16 | (buffer[1] & 0xFF) << 8 | (buffer[0] & 0xFF));
		//UID
		byte str_buf[] = new byte[32];
		
		b_in.read(str_buf,0,32);
		stat.uid = new String(str_buf);
		return stat;
	}
	
	/**
	 * 
	 * @param b_in
	 * @return
	 */
	protected assetTerminated readTermination(ByteArrayInputStream b_in) {
		assetTerminated at = new assetTerminated();
		// UID
		byte str_buf[] = new byte[32];
		b_in.read(str_buf, 0, 32);
		at.uid = new String(str_buf);
		return at;
	}
	
	/**
	 * 
	 * @param b_in
	 * @return
	 */
	protected radarDetection readDetection(ByteArrayInputStream b_in) {
		radarDetection rd = new radarDetection();
		// UID
		byte str_buf[] = new byte[32];
		b_in.read(str_buf, 0, 32);
		rd.uid = new String(str_buf);
		return rd;
	}
	
	public createStatus getStatus(String uid) {
		return map_status.get(uid);
	}
	
	public void closeServer() throws UnknownHostException, IOException {
		not_done = false;
	    s.leaveGroup(InetAddress.getByName(CREATE_GROUP));
	    s.close();
	}
}

